Creating aws_instance when there is no capacity should fail faster

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* aws_instance

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

```hcl
resource "aws_instance" "test" {
  ami                         = data.aws_ami.amzn-ami-minimal-hvm-ebs.id
  instance_type               = "t3.large"
  subnet_id                   = aws_subnet.test.id
  associate_public_ip_address = true
  tenancy                     = "dedicated"
  # pre-encoded base64 data
  user_data = "3dc39dda39be1205215e776bad998da361a5955d"
}

resource "aws_subnet" "test" {
  cidr_block              = "10.1.1.0/24"
  vpc_id                  = aws_vpc.test.id
  availability_zone       = data.aws_availability_zones.available.names[0]
  map_public_ip_on_launch = %[2]t

  tags = {
    Name = %[1]q
  }
}

resource "aws_vpc" "test" {
  cidr_block = "10.1.0.0/16"

  tags = {
    Name = %[1]q
  }
}

data "aws_availability_zones" "available" {
  state = "available"

  filter {
    name   = "opt-in-status"
    values = ["opt-in-not-required"]
  }
}
```

### Expected Behavior

AWS does not have any `t3.large` dedicated instances available in `us-west-2a`. The `aws_instance` should fail creation quickly.

```console
Error: Error launching source instance: InsufficientInstanceCapacity: We currently do not have sufficient t3.large capacity in the Availability Zone you requested (us-west-2a). Our system will be working on provisioning additional capacity. You can currently get t3.large capacity by not specifying an Availability Zone in your request or choosing us-west-2b, us-west-2c.
```

The error code is `Server.InsufficientInstanceCapacity`.

### Actual Behavior

The AWS SDK retried the request internally for approximately 50 minutes before returning the error. 

```console
    TestAccAWSInstance_vpc: resource_aws_instance_test.go:777: Step 1/2 error: terraform failed: exit status 1
        
        stderr:
        
        Error: Error launching source instance: InsufficientInstanceCapacity: We currently do not have sufficient t3.large capacity in the Availability Zone you requested (us-west-2a). Our system will be working on provisioning additional capacity. You can currently get t3.large capacity by not specifying an Availability Zone in your request or choosing us-west-2b, us-west-2c.
        	status code: 500, request id: cc98a6f3-deb6-4470-b7ca-c606f158467a
        
        
--- FAIL: TestAccAWSInstance_vpc (3101.11s)
```

--- a/.changelog/21293.txt
+++ b/.changelog/21293.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_instance: When encountering `InsufficientInstanceCapacity` errors, do not retry in order to fail faster, as this error is typically not resolvable in the near future
+```

--- a/internal/conns/config.go
+++ b/internal/conns/config.go
@@ -394,6 +394,12 @@ func (c *Config) ConfigureProvider(ctx context.Context, client *AWSClient) (*AWS
 			if tfawserr.ErrMessageContains(err, "VpnGatewayLimitExceeded", "maximum number of mutating objects has been reached") {
 				r.Retryable = aws.Bool(true)
 			}
+
+		case "RunInstances":
+			// `InsufficientInstanceCapacity` error has status code 500 and AWS SDK try retry this error by default.
+			if tfawserr.ErrCodeEquals(err, "InsufficientInstanceCapacity") {
+				r.Retryable = aws.Bool(false)
+			}
 		}
 	})
 

--- a/internal/service/ec2/ec2_ami_data_source_test.go
+++ b/internal/service/ec2/ec2_ami_data_source_test.go
@@ -110,7 +110,7 @@ func TestAccEC2AMIDataSource_windowsInstance(t *testing.T) {
 }
 
 func TestAccEC2AMIDataSource_instanceStore(t *testing.T) {
-	datasourceName := "data.aws_ami.test"
+	datasourceName := "data.aws_ami.amzn-ami-minimal-hvm-instance-store"
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:                 func() { acctest.PreCheck(t) },
@@ -205,7 +205,7 @@ func TestAccEC2AMIDataSource_gp3BlockDevice(t *testing.T) {
 // The data source is named 'amzn-ami-minimal-hvm-instance-store'.
 func testAccAMIDataSourceConfig_latestAmazonLinuxHVMInstanceStore() string {
 	return `
-data "aws_ami" "test" {
+data "aws_ami" "amzn-ami-minimal-hvm-instance-store" {
   most_recent = true
 
   filter {

--- a/internal/service/ec2/ec2_spot_fleet_request_test.go
+++ b/internal/service/ec2/ec2_spot_fleet_request_test.go
@@ -1665,7 +1665,7 @@ func TestAccEC2SpotFleetRequest_capacityRebalance(t *testing.T) {
 	})
 }
 
-func TestAccEC2SpotFleetRequest_withInstanceStoreAMI(t *testing.T) {
+func TestAccEC2SpotFleetRequest_instanceStoreAMI(t *testing.T) {
 	ctx := acctest.Context(t)
 	var config ec2.SpotFleetRequestConfig
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)

